hexl/experimental/misc/lr-mat-vec-mult.cpp

```diff
@@ -0,0 +1,154 @@
+// Copyright (C) 2020-2021 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+
+#include "hexl/experimental/misc/lr-mat-vec-mult.hpp"
+
+#include <cstring>
+
+#include "hexl/eltwise/eltwise-add-mod.hpp"
+#include "hexl/eltwise/eltwise-mult-mod.hpp"
+#include "hexl/number-theory/number-theory.hpp"
+#include "hexl/util/aligned-allocator.hpp"
+#include "hexl/util/check.hpp"
+#include "util/cpu-features.hpp"
+
+#define USE_ADDER_TREE
+
+namespace intel {
+namespace hexl {
+
+// operand1: num_weights x 2 x n x num_moduli [num_weights x |Ciphertext|].
+// operand2: num_weights x 2 x n x num_moduli [num_weights x
+// |Ciphertext/Plaintext|]. results: num_weights x 3 x n x num_moduli
+// [num_weights x {x[0].*y[0], x[0].*y[1]+x[1].*y[0], x[1].*y[1]} x num_moduli].
+// TODO(@fdiasmor): Ideally, the size of results can be optimized to [3 x n x
+// num_moduli].
+void LRMatrixVectorMultiply(uint64_t* result, const uint64_t* operand1,
+                            const uint64_t* operand2, uint64_t n,
+                            const uint64_t* moduli, uint64_t num_moduli,
+                            uint64_t num_weights) {
+  HEXL_CHECK(result != nullptr, "Require result != nullptr");
+  HEXL_CHECK(operand1 != nullptr, "Require operand1 != nullptr");
+  HEXL_CHECK(operand2 != nullptr, "Require operand2 != nullptr");
+  HEXL_CHECK(moduli != nullptr, "Require moduli != nullptr");
+  HEXL_CHECK(n != 0, "Require n != 0");
+  HEXL_CHECK(num_weights != 0, "Require n != 0");
+
+  // pointer increment to switch to a next polynomial
+  size_t poly_size = n * num_moduli;
+
+  // ciphertext increment to switch to the next ciphertext
+  size_t cipher_size = 2 * poly_size;
+
+  // ciphertext output increment to switch to the next output
+  size_t output_size = 3 * poly_size;
+
+  for (size_t r = 0; r < num_weights; r++) {
+    size_t next_output = r * output_size;
+    size_t next_poly_pair = r * cipher_size;
+    uint64_t* cipher2 = result + next_output;
+    const uint64_t* cipher0 = operand1 + next_poly_pair;
+    const uint64_t* cipher1 = operand2 + next_poly_pair;
+
+    for (size_t i = 0; i < num_moduli; i++) {
+      size_t i_times_n = i * n;
+      size_t poly0_offset = i_times_n;  // + tile_size * tile;
+      size_t poly1_offset = poly0_offset + poly_size;
+      size_t poly2_offset = poly0_offset + 2 * poly_size;
+
+      AlignedVector64<uint64_t> temp(n, 0);
```

Comment:
You may want to pull this initialization outside the for loop. I suppose CI will tell us, but it looks like you're writing up to `poly_size = n * num_moduli` coefficients to this vector size `n`?